### PR TITLE
hide left panel and map viz at SAM when map events occur

### PR DESCRIPTION
**Pull request by moontrip**
This addresses https://github.com/VEuPathDB/web-monorepo/issues/734. I think that in conjunction with left panel, it is better to close opened map Viz as well, thus this PR included that part as well.

Here is a video shot for a demo:

https://github.com/VEuPathDB/web-monorepo/assets/12802305/c9ce92c7-e119-4318-abfc-1a2c6d51bc60

### CODE REVIEW

**Comment by moontrip** [comment]
Hi @bobular your feedbacks at Slack are addressed so it would be great if you can check out.

File: packages/libs/components/src/map/MapVEuMap.tsx
Comment (dmfalke) [inline packages/libs/components/src/map/MapVEuMap.tsx:1]: Instead of the `setIsMapEvents` prop, I would rather see callback props for each event. The current approach requires state to track when an event is triggered, and it also doesn't indicate which event was triggered.

The props could be as simple as

```typescript
interface MapVEuMapProps {
  // ...
  onClick: () => void;
  onDrag: () => void;
  onZoom: () => void;
}
```

File: packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx
Comment (dmfalke) [inline packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx:1]: Following on to what I said above, the callback used for each event handler should just call `appState.setIsSidePanelExpanded(false)`.

**Review by dmfalke — CHANGES_REQUESTED** [review]
@moontrip I made some comments recommending a different approach.

The upshot is that `MapVEuMap` should take callback props for each event, and `MapAnalysis` should pass a callback that updates the appState so that `isSidePanelExpanded = false`.

**Comment by moontrip** [comment]
> @moontrip I made some comments recommending a different approach.
> 
> The upshot is that `MapVEuMap` should take callback props for each event, and `MapAnalysis` should pass a callback that updates the appState so that `isSidePanelExpanded = false`.

Hi @dmfalke 👋 Thank you for your suggestions. I will test them next week 👍 Have a great weekend!

**Comment by dmfalke** [comment]
> Hi @dmfalke 👋 Thank you for your suggestions. I will test them next week 👍 Have a great weekend!

🤘

File: packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx
Comment (bobular) [reply packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx:1]: I agree with Dave!

**Comment by moontrip** [comment]
Hi @dmfalke and @bobular 

I have addressed your feedbacks with some changes. One question I have is that actually we do not necessarily pass three props such as onDrag, onZoom, and onClick because a map event (drag, zoom, or click) is detected at corresponding MapVEuMapEvents in the MapVEuMap component. So, perhaps we can simply pass single prop such as setIsSidePanelExpanded(), to MapVEuMapEvents /MapVEuMap?

**Comment by dmfalke** [comment]
> Hi @dmfalke and @bobular
> 
> I have addressed your feedbacks with some changes. One question I have is that actually we do not necessarily pass three props such as onDrag, onZoom, and onClick because a map event (drag, zoom, or click) is detected at corresponding MapVEuMapEvents in the MapVEuMap component. So, perhaps we can simply pass single prop such as setIsSidePanelExpanded(), to MapVEuMapEvents /MapVEuMap?

I think we should keep the three props. That gives us flexibility in the future, in case we want to change which events cause the side panel to be collapsed.

File: packages/libs/components/src/map/MapVEuMap.tsx [hunk 1]
```diff
@@ -303,6 +313,9 @@ interface MapVEuMapEventsProps {
   onBoundsChanged: (bondsViewport: BoundsViewport) => void;
   onBaseLayerChanged?: (newBaseLayer: BaseLayerChoice) => void;
   setSelectedMarkers?: React.Dispatch<React.SetStateAction<string[]>>;
+  onMapClick?: (value: boolean) => void;
+  onMapDrag?: (value: boolean) => void;
+  onMapZoom?: (value: boolean) => void;
```
Comment (dmfalke) [inline packages/libs/components/src/map/MapVEuMap.tsx:318]: Let's keep these generic and call them with any arguments. In other uses cases, it wouldn't be clear what the boolean `value` means.

```suggestion
  onMapClick?: () => void;
  onMapDrag?: () => void;
  onMapZoom?: () => void;
```

File: packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx [hunk 2]
```diff
@@ -900,6 +898,9 @@ function MapAnalysisImpl(props: ImplProps) {
                     defaultViewport={defaultViewport}
                     // for multiple markers cancelation of selection only
                     setSelectedMarkers={setSelectedMarkers}
+                    onMapClick={setIsSidePanelExpanded}
+                    onMapDrag={setIsSidePanelExpanded}
+                    onMapZoom={setIsSidePanelExpanded}
```
Comment (dmfalke) [inline packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx:903]: Instead of passing the in the `useState` setter function, use a wrapper function that calls it:

```ts
const closePanel = useCallback(() => setIsSidePanelExpanded(false), []);
```

```suggestion
                    onMapClick={closePanel}
                    onMapDrag={closePanel}
                    onMapZoom={closePanel}
```

**Review by dmfalke — CHANGES_REQUESTED** [review]

**Comment by moontrip** [comment]
> > Hi @dmfalke and @bobular
> > I have addressed your feedbacks with some changes. One question I have is that actually we do not necessarily pass three props such as onDrag, onZoom, and onClick because a map event (drag, zoom, or click) is detected at corresponding MapVEuMapEvents in the MapVEuMap component. So, perhaps we can simply pass single prop such as setIsSidePanelExpanded(), to MapVEuMapEvents /MapVEuMap?
> 
> I think we should keep the three props. That gives us flexibility in the future, in case we want to change which events cause the side panel to be collapsed.

Thank @dmfalke for your comment and suggestions. I applied them to the new commit 👍

File: packages/libs/components/src/map/MapVEuMap.tsx [hunk 3]
```diff
@@ -154,6 +154,10 @@ export interface MapVEuMapProps {
   /* selectedMarkers setState (for on-click reset) **/
   setSelectedMarkers?: React.Dispatch<React.SetStateAction<string[]>>;
   children?: React.ReactNode;
+  /* close left-side panel when map events happen */
```
Comment (bobular) [inline packages/libs/components/src/map/MapVEuMap.tsx:157]: We shouldn't be documenting behaviour of the map code here. The callbacks can do *anything*!

File: packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx [hunk 4]
```diff
@@ -871,9 +872,9 @@ function MapAnalysisImpl(props: ImplProps) {
                   <MapSidePanel
                     isExpanded={appState.isSidePanelExpanded}
                     isUserLoggedIn={userLoggedIn}
-                    onToggleIsExpanded={() =>
-                      setIsSidePanelExpanded(!appState.isSidePanelExpanded)
-                    }
+                    onToggleIsExpanded={() => {
+                      setIsSidePanelExpanded(!appState.isSidePanelExpanded);
+                    }}
```
Comment (bobular) [inline packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx:877]:
```suggestion
                    onToggleIsExpanded={() =>
                      setIsSidePanelExpanded(!appState.isSidePanelExpanded)
                    }
```

I'd go back to the original. Maybe you added logging here but best to not make unnecessary changes.

File: packages/libs/components/src/map/MapVEuMap.tsx [hunk 5]
```diff
@@ -343,13 +362,19 @@ function MapVEuMapEvents(props: MapVEuMapEventsProps) {
         zoomLevel: mapEvents.getZoom(),
       };
       onBoundsChanged(boundsViewport);
+
+      // close left-side panel when drag/panning event happens
```
Comment (bobular) [inline packages/libs/components/src/map/MapVEuMap.tsx:366]: Same comment about this comment (and the ones above and below). Best to remove.

File: packages/libs/components/src/map/MapVEuMap.tsx (hunk 3 above)
Comment (moontrip) [reply packages/libs/components/src/map/MapVEuMap.tsx]: Thank you for your review and suggestions @bobular I got your point and remove it.

File: packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx (hunk 4 above)
Comment (moontrip) [reply packages/libs/eda/src/lib/map/analysis/MapAnalysis.tsx]: complied!

File: packages/libs/components/src/map/MapVEuMap.tsx (hunk 5 above)
Comment (moontrip) [reply packages/libs/components/src/map/MapVEuMap.tsx]: all removed!

**Review by bobular — APPROVED** [review]
Looks great, thanks!

**Review by dmfalke — APPROVED** [review]
Looks great!

**Comment by moontrip** [comment]
Thanks @bobular and @dmfalke for your reviews and feedbacks 👍
